File: code/tcp-udp/udp_server.py
import socket


def main():
    host = '127.0.0.1'
    port = 8888

    server = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    server.bind((host, port))

    print(f'UDP Server 启动, 监听端口: {port}')
    print('UDP 是无连接的, 不需要 3 次握手')

    data, addr = server.recvfrom(1024)
    msg = data.decode()
    print(f'收到: {msg} 来自: {addr[0]}:{addr[1]}')

    resp = f'ECHO: {msg}'.encode()
    server.sendto(resp, addr)

    server.close()


if __name__ == '__main__':
    main()

File: code/tcp-udp/UdpServer.java
import java.net.*;

/**
 * UDP 服务器示例
 * 演示 UDP 无连接通信
 *
 * 运行方式:
 *   javac UdpServer.java && java UdpServer
 */
public class UdpServer {

    public static void main(String[] args) throws Exception {
        int port = 8888;
        DatagramSocket socket = new DatagramSocket(port);
        System.out.println("UDP Server 启动, 监听端口: " + port);
        System.out.println("UDP 是无连接的, 不需要 3 次握手");

        byte[] buf = new byte[1024];
        DatagramPacket packet = new DatagramPacket(buf, buf.length);

        socket.receive(packet);
        String msg = new String(packet.getData(), 0, packet.getLength());
        System.out.println("收到: " + msg + " 来自: " + packet.getAddress());

        // 回复
        byte[] resp = ("ECHO: " + msg).getBytes();
        DatagramPacket respPacket = new DatagramPacket(
                resp, resp.length, packet.getAddress(), packet.getPort());
        socket.send(respPacket);

        socket.close();
    }
}
